src/storage/garbage_collector.cpp

```diff
@@ -86,80 +117,284 @@ uint32_t GarbageCollector::ProcessUnlinkQueue() {
+  *curr_ptr = compacted_undo_record;
+}
+
+void GarbageCollector::ReadUndoRecord(UndoRecord *start_record, UndoRecord *next, uint32_t *interval_length_ptr) {
+  // Already have a base undo record. Apply this undo record on top of that
+  switch (next->Type()) {
+    case DeltaRecordType::UPDATE:
+      // Update the interval length
+      (*interval_length_ptr)++;
+      // Process the Attributes to determine the attributes required for the compacted undo record
+      ProcessUndoRecordAttributes(next);
+      // Mark the Record as unlinked
+      UnlinkUndoRecordVersion(next);
+      break;
+    case DeltaRecordType::INSERT:
+      // Compacting here. So unlink start_record's next
+      UnlinkUndoRecordVersion(start_record->Next());
+      EndCompaction(interval_length_ptr);
+      // Insert undo record can be GC'd so this tuple is not visible
+      // Set start_record to point to Insert's next undo record
+      SwapwithSafeAbort(start_record, next);
+      break;
+    case DeltaRecordType::DELETE: {
+    }
+  }
+}
+
+void GarbageCollector::EndCompaction(uint32_t *interval_length_ptr) { *interval_length_ptr = 0; }
+
+void GarbageCollector::ProcessUndoRecordAttributes(UndoRecord *const undo_record) {
+  for (uint16_t i = 0; i < undo_record->Delta()->NumColumns(); i++) {
+    col_id_t col_id = undo_record->Delta()->ColumnIds()[i];
+
+    // Add col to the col_set_
+    col_set_.insert(col_id);
+  }
+}
+
+UndoRecord *GarbageCollector::CreateUndoRecord(UndoRecord *const start_record, UndoRecord *const end_record) {
+  UndoRecord *curr = start_record->Next();
+  UndoRecord *first_compacted_record = start_record->Next();
+  DataTable *table = first_compacted_record->Table();
+  const TupleAccessStrategy &accessor = table->accessor_;
+
+  // Initialize the base Undo record
+  UndoRecord *base_undo_record =
+      InitializeUndoRecord(first_compacted_record->Timestamp().load(), first_compacted_record->Slot(), table);
+
+  // Apply all the undo records which have to be compacted
+  while (curr != end_record) {
+    StorageUtil::ApplyDelta(accessor.GetBlockLayout(), *(curr->Delta()), base_undo_record->Delta());
+    curr = curr->Next();
+  }
+  return base_undo_record;
+}
+
+UndoRecord *GarbageCollector::InitializeUndoRecord(const transaction::timestamp_t timestamp, const TupleSlot slot,
+                                                   DataTable *const table) {
+  const TupleAccessStrategy &accessor = table->accessor_;
+  const BlockLayout &layout = accessor.GetBlockLayout();
+
+  // Initialize the projected row with the set of columns we have seen in the first pass
+  std::vector<col_id_t> col_id_list(col_set_.begin(), col_set_.end());
+  auto init = storage::ProjectedRowInitializer::CreateProjectedRowInitializer(layout, col_id_list);
+
+  // Get new entry for the undo record from the buffer
+  uint32_t size = static_cast<uint32_t>(sizeof(UndoRecord)) + init.ProjectedRowSize();
+  byte *head = delta_record_compaction_buffer_->NewEntry(size);
```

Comment:
Might be good to assert head here

Reply:
Can you clarify what you mean?
Do you mean something like `ASSERT(head != nullptr)`?

Reply:
Yes, or have a control flow for handling the nullptr case

Reply:
Added check in the next line.